### PR TITLE
Add Youtube repositories

### PR DESCRIPTION
**Pull request by sheetal-g9**
First steps for issues #119 
Changes:Add playlist and video repositories ,fetching data with api and storing them in playlist and video models.

### CODE REVIEW

**Review by dr0pdb — CHANGES_REQUESTED** [review]
Go through the codacy issues and fix them. Also look at why the build is failing and fix it.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/VideoRepository.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                    videos.setValue(youtubeVideos);
+                }
+
+                @Override
+                public void onError(Throwable e) {
+                    e.printStackTrace();
+                }
+
+                @Override
+                public void onComplete() {
```
[the remark this answers is not in the document]
Comment (BVShe001) [reply app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/VideoRepository.java]: Observer requires this method to be implemented.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        mService = new com.google.api.services.youtube.YouTube.Builder(
+            transport, jsonFactory, googleAccountCredential)
+            .setApplicationName("IITP-Connect")
+            .build();
+        io.reactivex.Observable.just(channelName).map(s -> getFromApi(mService))
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(new Observer<List<YoutubePlaylist>>() {
+                @Override
+                public void onSubscribe(Disposable d) {
```
[the remark this answers is not in the document]
Comment (BVShe001) [reply app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java]: Observer requires this method to be implemented.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                    playlists.setValue(youtubePlaylists);
+                }
+
+                @Override
+                public void onError(Throwable e) {
+                    e.printStackTrace();
+                }
+
+                @Override
+                public void onComplete() {
```
[the remark this answers is not in the document]
Comment (BVShe001) [reply app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java]: Observer requires this method to be implemented.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/VideoRepository.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            transport, jsonFactory, googleAccountCredential)
+            .setApplicationName("IITP-Connect")
+            .build();
+
+        io.reactivex.Observable.just(playlist_id).map(s -> getFromApi(mService))
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(new Observer<List<YoutubeVideo>>() {
+                @Override
+                public void onSubscribe(Disposable d) {
```
[the remark this answers is not in the document]
Comment (BVShe001) [reply app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/VideoRepository.java]: Observer requires this method to be implemented.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (srv-twry) [reply app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java]: For empty body add a comment such as `// nothing here`.

**Comment by dr0pdb** [comment]
@BVShe001 I know that those empty methods are required but Codacy is suggesting you to **document** it not remove it. Add a comment like `nothing here` inside the method body.

**Comment by sheetal-g9** [comment]
Build is failing because it is not getting api key which I have abstracted as in :
[ how to hide](https://medium.com/code-better/hiding-api-keys-from-your-android-repository-b23f5598b906)
It is building perfectly in my Android Studio.

**Comment by dr0pdb** [comment]
@BVShe001 
1. If we're going to let the build fail now and merge this one then all the future builds will fail. Then what's the point of these CI's?
2. Search for how to setup demo api-key on Travis so that the build doesn't fail.

File: .travis.yml [hunk 5]
```diff
@@ -1,27 +1,30 @@
...
 android:
   components:
-    - tools
-    - android-27
-    - build-tools-27.0.3
-    - platform-tools
+  - tools
+  - android-27
+  - build-tools-27.0.3
+  - platform-tools
```
Comment (dr0pdb) [inline .travis.yml]: What have you done here? Don't make unnecessary changes.

**Review by dr0pdb — CHANGES_REQUESTED** [review]

File: app/build.gradle [hunk 6]
```diff
@@ -110,7 +110,7 @@ dependencies {
     implementation "com.squareup.retrofit2:retrofit:$rootProject.retrofitVersion"
     implementation "com.squareup.retrofit2:converter-jackson:$rootProject.retrofitVersion"
     implementation "com.squareup.retrofit2:adapter-rxjava2:$rootProject.retrofitVersion"
-    implementation 'com.squareup.okhttp3:logging-interceptor:3.12.0'
+    implementation 'com.squareup.okhttp3:logging-interceptor:3.11.0'
```
Comment (dr0pdb) [inline app/build.gradle]: Why decrease the version?

File: app/build.gradle [hunk 7]
```diff
@@ -171,9 +171,27 @@ dependencies {
         exclude group: 'com.android.support', module: 'support-annotations'
     })
     testImplementation 'junit:junit:4.12'
-    testImplementation 'org.mockito:mockito-inline:2.23.4'
+    testImplementation 'org.mockito:mockito-inline:2.23.0'
```
Comment (dr0pdb) [inline app/build.gradle]: Why decrease it?

File: app/build.gradle [hunk 8]
```diff
@@ -77,9 +80,6 @@ check {
 dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
 
-    androidTestImplementation('com.android.support.test.espresso:espresso-core:2.2.2', {
-        exclude group: 'com.android.support', module: 'support-annotations'
-    })
```
Comment (dr0pdb) [inline app/build.gradle:82]: Why remove it?

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/models/YoutubePlaylist.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        return playlist_id;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public long getItemCount() {
+        return itemCount;
+    }
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/models/YoutubePlaylist.java]: Please take a look at other model classes. These getters, setters, constructors can easily be done using annotations such as `@data`, `@allargsconstructor` etc. Use them.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/models/YoutubeVideo.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        return title;
+    }
+
+    public String getPublishedAt() {
+        return publishedAt.toString().substring(0, 10);
+    }
+
+    public String getVideoId() {
+        return video_id;
+    }
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/models/YoutubeVideo.java]: Check other models.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+import javax.inject.Inject;
+
+import io.reactivex.Observer;
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.schedulers.Schedulers;
+
+public class PlaylistRepository {
+    public GoogleAccountCredential googleAccountCredential;
+    private String channelName = "GoogleDevelopers";
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java]: What is this?

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    public MutableLiveData<List<YoutubePlaylist>> getPlaylists() {
+        return this.playlists;
+    }
+
+    public void getDataFromAPI() {
+        HttpTransport transport = AndroidHttp.newCompatibleTransport();
+        JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+        service = new YouTube.Builder(
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/PlaylistRepository.java]: Use Dagger for it. Create an instance of YouTube builder and inject it in the Repository as a dependency just like `GoogleAccountCredential`.

File: app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/VideoRepository.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    public void setPlaylistId(String playlist_id) {
+        this.playlist_id = playlist_id;
+    }
+
+    public void getDataFromAPI() {
+        HttpTransport transport = AndroidHttp.newCompatibleTransport();
+        JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+        service = new YouTube.Builder(
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/core/youtube/repositories/VideoRepository.java]: Use Dagger.

**Review by dr0pdb — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 8 above)
Comment (sheetal-g9) [reply app/build.gradle:82]: It is redundant.Same dependency is used at line :170 with a newer version 3.0.1

File: app/src/main/java/com/iitp/njack/iitp_connect/common/di/modules/YoutubeModule.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+import dagger.Module;
+import dagger.Provides;
+
+@Module
+public class YoutubeModule {
+    @Provides
+    @Singleton
+    PlaylistRepository providesPlaylistRepository() {
+        return new PlaylistRepository();
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/common/di/modules/YoutubeModule.java]: Don't use DI for providing repository. Revert this back to what it was.

File: app/src/main/java/com/iitp/njack/iitp_connect/common/di/modules/YoutubeModule.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        String[] SCOPES = {YouTubeScopes.YOUTUBE_READONLY};
+        return GoogleAccountCredential.usingOAuth2(
+            context, Arrays.asList(SCOPES))
+            .setBackOff(new ExponentialBackOff());
+    }
+
+    @Provides
+    @Singleton
+    VideoRepository providesVideoRepository() {
+        return new VideoRepository();
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/common/di/modules/YoutubeModule.java]: Same here.

File: app/src/main/java/com/iitp/njack/iitp_connect/common/di/modules/YoutubeModule.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Provides
+    @Singleton
+    HttpTransport providesHttpTransport(){
+        return AndroidHttp.newCompatibleTransport();
+    }
```
Comment (dr0pdb) [inline app/src/main/java/com/iitp/njack/iitp_connect/common/di/modules/YoutubeModule.java:39]: Add an extra line!

**Review by dr0pdb — CHANGES_REQUESTED** [review]

**Pull request closed without merging**